JIT - handle multiple expressions (as returned from CSE)

mdewing opened the pull request:
The `expr` argument to llvm_callable can either be a single expression, or a structure that matches the return value from `cse`.

The expected structure is a tuple or list with two items - first item is a list of (Symbol, expr) pairs, and the second item is a list with one item - the final expression to be returned.
(The cse module can handle multiple expressions, and in that case will put multiple items in that second list.  This PR will only process and return the value from the first expression on that list)

This code also assumes the list of (Symbol, expr) is topologically sorted.

Example usage

``` python
from sympy import *
import sympy.printing.llvmjitcode as jit
from sympy.abc import a,x,y,z

e = x*x + y*y  + exp(-x*x- y*y)
f = x*x + exp(-x*x- y*y)
print('e = ',e)
e2 = cse(e)
print('e2 = ',e2)

j1 = jit.llvm_callable([x,y],e)
print("original expression = ",j1(1.0, 2.0))

j2 = jit.llvm_callable([x,y],e2)
print("after cse = ",j2(1.0, 2.0))

```

mdewing committed:
JIT compile structures from cse.

Handle expressions with temporary internal values, as returned from
the cse module.

asmeurer commented:
> This PR will only process and return the value from the first expression on that list

It might be better to just raise a NotImplementedError. Only processing the first item doesn't sound like expected behavior.

mdewing committed:
JIT - raise exception in multiple expressions case

The cse module supports returning multiple expressions, but the JIT
code only supports one.  Raise an exception in this case.
Also add an exception when unknown symbols are encountered in expressions
(most likely they were forgotten in the argument list).

markdewing commented:
Here's an example using 6-D integration to approximate the ground state energy of a helium atom.  It also shows the use of CSE to improve the performance of the callback (and the integration) - 195 seconds vs. 140 seconds with CSE.

https://github.com/markdewing/next_steps_in_programming/blob/master/examples/integration/Helium%20atom.ipynb

moorepants commented:
Great example! Feel free to submit that for our examples directory.

asmeurer commented on sympy/printing/tests/test_llvmjit.py:
```diff
@@ -44,10 +45,138 @@ def test_func():
+    e3 = sympy.cse([e1, e2])
+    try:
+        g.llvm_callable([a, b], e3)
+    except NotImplementedError:
```
You should use the `raises` function here.

mdewing committed:
LLVM JIT tests - use 'raises' function.

mdewing committed:
Merge remote-tracking branch 'upstream/master' into cse_jit

LLVM JIT - uses 'raises' function in tests.

mdewing committed:
Merge remote-tracking branch 'upstream/master' into cse_jit

mdewing committed:
JIT - handle multiple expressions.

The cse module can process multiple expressions.  The compiled function
returns multiple values as a tuple.
For the cubature callback, integrating multiple similar expressions at the same
time can save time (compared to separate calls to the integrator).
Multiple expressions are not supported with the scipy.integrate callback.

markdewing commented:
Ping @asmeurer @moorepants 
Could this PR get merged?
